src/Common/src/Microsoft/Win32/SafeHandles/SafeHandleCache.cs

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace Microsoft.Win32.SafeHandles
+{
+    /// <summary>Provides a cache for special instances of SafeHandles.</summary>
+    /// <typeparam name="T">Specifies the type of SafeHandle.  The type must have a parameterless constructor.</typeparam>
+    internal static class SafeHandleCache<T> where T : SafeHandle
+    {
+        private static T s_invalidHandle;
+
+        /// <summary>
+        /// Gets a cached, invalid handle.  As the instance is cached, it should either never be Disposed
+        /// or it should override <see cref="SafeHandle.Dispose(bool)"/> to prevent disposal when the
+        /// instance is the <see cref="InvalidHandle"/>.
+        /// </summary>
+        internal static T InvalidHandle
+        {
+            get
+            {
+                T currentHandle = Volatile.Read(ref s_invalidHandle);
+                if (currentHandle == null)
+                {
+                    T newHandle = Activator.CreateInstance<T>();
+                    currentHandle = Interlocked.CompareExchange(ref s_invalidHandle, newHandle, null);
+                    if (currentHandle == null)
+                    {
+                        GC.SuppressFinalize(newHandle);
+                        currentHandle = newHandle;
+                    }
+                    else
+                    {
+                        newHandle.Dispose();
+                    }
+                }
+                Debug.Assert(currentHandle.IsInvalid);
+                return currentHandle;
+            }
+        }
+
+        /// <summary>Gets whether the specified handle is <see cref="InvalidHandle"/>.</summary>
+        /// <param name="handle">The handle to compare.</param>
+        /// <returns>true if <paramref name="handle"/> is <see cref="InvalidHandle"/>; otherwise, false.</returns>
+        internal static bool IsCachedInvalidHandle(SafeHandle handle)
+        {
+            Debug.Assert(handle != null);
+            bool isCachedInvalidHandle = ReferenceEquals(handle, Volatile.Read(ref s_invalidHandle));
```

Comment:
:bulb: On X86, this read definitely does not need to be volatile because the `CompareExchange` instruction which sets the field uses the **lock** prefix, and thus has a full memory barrier. I don't know of any system which implements `CompareExchange` in any other way.